### PR TITLE
Support exclude_monikers and replace_monikers

### PR DESCRIPTION
**Pull request by live1206**
[AB#268902](https://dev.azure.com/ceapex/Engineering/_workitems/edit/268902)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6267)

### CODE REVIEW

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 1]
```diff
@@ -112,7 +130,7 @@ public int GetMonikerOrder(string moniker)
                 if (!_config.SkipMonikerValidation && configMonikerRange.Value is null)
                 {
                     errors.Add(Errors.Versioning.MonikerRangeUndefined(metadata.MonikerRange.Source));
-                    return (errors, configMonikers);
+                    return (errors, configMonikers.Except(excludeMonikers));
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:133]: should `exclude_monikers` be processed _before_ checking intersection? Aka, should we report error when a file produces an empty moniker list due to `exclude_monikers`?

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.Graph;
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:8]: revert?

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 3]
```diff
@@ -104,45 +105,84 @@ public int GetMonikerOrder(string moniker)
...
+                        GetMonikerIntersection(metadata, configMonikerRange, configMonikers, fileMonikers, _config.SkipMonikerValidation);
+                    errors.AddIfNotNull(intersectionError);
+                    return (errors, intersection);
                 }
+            }
+
+            return (errors, configMonikers);
+        }
+
+        private bool TryGetFileMonikers(UserMetadata metadata, out List<Error> errors, out MonikerList fileMonikers)
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:136]:
```suggestion
        private bool TryGetFileMonikers(UserMetadata metadata, out List<Error> errors, [NotNullWhen(true)] out MonikerList? fileMonikers)
```

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 4]
```diff
@@ -104,45 +105,84 @@ public int GetMonikerOrder(string moniker)
...
+                    // Moniker range not defined in docfx.yml/docfx.json,
+                    // user should not define it in file metadata
+                    if (!_config.SkipMonikerValidation && configMonikerRange.Value is null)
+                    {
+                        errors.Add(Errors.Versioning.MonikerRangeUndefined(metadata.MonikerRange.Source));
+                        return (errors, configMonikers);
+                    }
+                }
+
+                if (fileMonikers != null)
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:124]:
```suggestion
                ~~~if (fileMonikers != null)~~~
```

**Comment by live1206** [comment]
> ⚠[sql-docs-pr](https://github.com/MicrosoftDocs/sql-docs-pr)(00:01:04.2364464, 0 more diff)
> ```diff
> diff --git "a/D:/docfx.TestData\\regression-test.sql-docs-pr\\output/docs/.errors.log" "b/D:/docfx.TestData\\regression-test.sql-docs-pr\\.temp/docs/.errors.log"
> index 4ae0431..10a43e6 100644
> --- "a/D:/docfx.TestData\\regression-test.sql-docs-pr\\output/docs/.errors.log"
> +++ "b/D:/docfx.TestData\\regression-test.sql-docs-pr\\.temp/docs/.errors.log"
> @@ -407519,7 +407519,7 @@
>    "message_severity": "suggestion",
>    "log_item_type": "user",
>    "code": "toc-entry-duplicated",
> -  "message": "Article is referenced in more than one TOC node: data-virtualization.md, upgrade-dbcompat-using-qta.md, reorganize-and-rebuild-indexes.md, tutorial-sql-server-containers-kubernetes.md, query-and-text-editors-sql-server-management-studio.md, xml-editor-sql-server-management-studio.md, color-coding-in-query-editors.md, configure-editors-sql-server-management-studio.md, associate-file-extensions-to-a-code-editor.md, change-font-color-size-and-style.md, use-the-properties-window-in-management-studio.md, open-an-editor-sql-server-management-studio.md, manage-the-editor-and-view-mode.md, manage-code-formatting.md, navigate-code-and-text.md, sql-server-management-objects-smo-programming-guide.md. Each article can only be referenced once in a TOC.",
> +  "message": "Article is referenced in more than one TOC node: data-virtualization.md, reorganize-and-rebuild-indexes.md, upgrade-dbcompat-using-qta.md, tutorial-sql-server-containers-kubernetes.md, query-and-text-editors-sql-server-management-studio.md, xml-editor-sql-server-management-studio.md, color-coding-in-query-editors.md, configure-editors-sql-server-management-studio.md, associate-file-extensions-to-a-code-editor.md, change-font-color-size-and-style.md, use-the-properties-window-in-management-studio.md, open-an-editor-sql-server-management-studio.md, manage-the-editor-and-view-mode.md, manage-code-formatting.md, navigate-code-and-text.md, sql-server-management-objects-smo-programming-guide.md. Each article can only be referenced once in a TOC.",
>    "file": "toc.yml",
>    "line": 0,
>    "end_line": 0,
> @@ -407552,8 +407552,8 @@
>    "message_severity": "suggestion",
>    "log_item_type": "user",
>    "code": "toc-entry-duplicated",
> -  "message": "Article is referenced in more than one TOC node: manage-the-database-engine-services.md, server-network-configuration.md, client-network-configuration.md, force-a-wsfc-cluster-to-start-without-a-quorum.md, use-the-new-availability-group-dialog-box-sql-server-management-studio.md. Each article can only be referenced once in a TOC.",
> -  "file": "2014/database-engine/toc.yml",
> +  "message": "Article is referenced in more than one TOC node: usage-data-collection.md. Each article can only be referenced once in a TOC.",
> +  "file": "azure-data-studio/toc.yml",
>    "line": 0,
>    "end_line": 0,
>    "column": 0,
> @@ -407563,8 +407563,8 @@
>    "message_severity": "suggestion",
>    "log_item_type": "user",
>    "code": "toc-entry-duplicated",
> -  "message": "Article is referenced in more than one TOC node: usage-data-collection.md. Each article can only be referenced once in a TOC.",
> -  "file": "azure-data-studio/toc.yml",
> +  "message": "Article is referenced in more than one TOC node: use-the-new-availability-group-dialog-box-sql-server-management-studio.md, force-a-wsfc-cluster-to-start-without-a-quorum.md, manage-the-database-engine-services.md, server-network-configuration.md, client-network-configuration.md. Each article can only be referenced once in a TOC.",
> +  "file": "2014/database-engine/toc.yml",
>    "line": 0,
>    "end_line": 0,
>    "column": 0,
> ```

@OsmondJiang The error message of `toc-entry-duplicated` is not stable

**Pull request merged**